coap/api/transport.go

```diff
@@ -66,41 +67,53 @@ func MakeCoAPHandler(svc coap.Service, l *slog.Logger) mux.HandlerFunc {
 	return handler
 }
 
-func sendResp(w mux.ResponseWriter, resp *message.Message) {
-	if err := w.Client().WriteMessage(resp); err != nil {
-		logger.Warn(fmt.Sprintf("Can't set response: %s", err))
+func sendResp(w mux.ResponseWriter, resp *message.Message, isObs bool) {
+	pm := w.Conn().AcquireMessage(w.Conn().Context())
+	defer w.Conn().ReleaseMessage(pm)
+	pm.SetCode(resp.Code)
+	pm.SetToken(resp.Token)
+	pm.SetBody(bytes.NewReader(resp.Payload))
+	for _, opt := range resp.Options {
+		pm.SetOptionBytes(opt.ID, opt.Value)
+	}
+	if !isObs {
+		if err := w.Conn().WriteMessage(pm); err != nil {
+			logger.Warn(fmt.Sprintf("Can't set response: %s", err))
+		}
 	}
 }
 
 func handler(w mux.ResponseWriter, m *mux.Message) {
 	resp := message.Message{
 		Code:    codes.Content,
-		Token:   m.Token,
-		Context: m.Context,
+		Token:   m.Token(),
 		Options: make(message.Options, 0, 16),
 	}
-	defer sendResp(w, &resp)
 	msg, err := decodeMessage(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error decoding message: %s", err))
 		resp.Code = codes.BadRequest
+		sendResp(w, &resp, false)
 		return
 	}
 	key, err := parseKey(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error parsing auth: %s", err))
 		resp.Code = codes.Unauthorized
+		sendResp(w, &resp, false)
 		return
 	}
-	switch m.Code {
+	var isObs bool = false
+	switch m.Code() {
 	case codes.GET:
-		err = handleGet(m.Context, m, w.Client(), msg, key)
+		isObs, err = handleGet(context.Background(), m, w, msg, key)
 	case codes.POST:
 		resp.Code = codes.Created
-		err = service.Publish(m.Context, key, msg)
+		err = service.Publish(m.Context(), key, msg)
 	default:
 		err = svcerr.ErrNotFound
 	}
```

Comment:
```suggestion
func handler(w mux.ResponseWriter, m *mux.Message) {
	isObs := false
	resp := message.Message{
		Code:    codes.Content,
		Token:   m.Token(),
		Options: make(message.Options, 0, 16),
	}
	defer sendResp(w, &resp, isObs)
	msg, err := decodeMessage(m)
	if err != nil {
		logger.Warn(fmt.Sprintf("Error decoding message: %s", err))
		resp.Code = codes.BadRequest
		return
	}
	key, err := parseKey(m)
	if err != nil {
		logger.Warn(fmt.Sprintf("Error parsing auth: %s", err))
		resp.Code = codes.Unauthorized
		return
	}
	switch m.Code() {
	case codes.GET:
		isObs, err = handleGet(context.Background(), m, w, msg, key)
	case codes.POST:
		resp.Code = codes.Created
		err = service.Publish(m.Context(), key, msg)
	default:
		err = svcerr.ErrNotFound
	}
```

Reply:
Another way, 

```suggestion
func sendResp(w mux.ResponseWriter, resp *pool.Message) {
	if err := w.Conn().WriteMessage(resp); err != nil {
		logger.Warn(fmt.Sprintf("Can't set response: %s", err))
	}
}

func handler(w mux.ResponseWriter, m *mux.Message) {
	resp := pool.NewMessage(w.Conn().Context())
	resp.SetToken(m.Token())
	defer sendResp(w, resp)

	msg, err := decodeMessage(m)
	if err != nil {
		logger.Warn(fmt.Sprintf("Error decoding message: %s", err))
		resp.SetCode(codes.BadRequest)
		return
	}
	key, err := parseKey(m)
	if err != nil {
		logger.Warn(fmt.Sprintf("Error parsing auth: %s", err))
		resp.SetCode(codes.Unauthorized)
		return
	}
	switch m.Code() {
	case codes.GET:
		isObs := false
		isObs, err = handleGet(context.Background(), m, w, msg, key)
		if isObs && err != nil {
			resp.SetCode(codes.Content)
		}
	case codes.POST:
		resp.SetCode(codes.Created)
		err = service.Publish(m.Context(), key, msg)
	default:
		err = svcerr.ErrNotFound
	}
	if err != nil {
		switch {
		case err == errBadOptions:
			resp.SetCode(codes.BadOption)
		case err == svcerr.ErrNotFound:
			resp.SetCode(codes.NotFound)
		case errors.Contains(err, svcerr.ErrAuthorization),
			errors.Contains(err, svcerr.ErrAuthentication):
			resp.SetCode(codes.Unauthorized)
		default:
			resp.SetCode(codes.InternalServerError)
		}
	}
}
```

Reply:
I had tried the first and it results in the linter flagging an ineffectual assignment on line 108, let me try the second one